ouroboros-network/src/Ouroboros/Network/Subscription/Common.hs

```diff
@@ -0,0 +1,383 @@
+{-# LANGUAGE CPP                 #-}
+{-# LANGUAGE DataKinds           #-}
+{-# LANGUAGE RankNTypes          #-}
+{-# LANGUAGE RecordWildCards     #-}
+{-# LANGUAGE ScopedTypeVariables #-}
+
+module Ouroboros.Network.Subscription.Common
+    ( subscribeTo
+    , sockAddrFamily
+    , minConnectionAttemptDelay
+    , ipSubscriptionWorker
+    , subscriptionWorker
+    , SubscriberError (..)
+    , SubscriptionTrace
+    , IPSubscriptionTarget (..)
+    ) where
+
+import           Control.Concurrent hiding (threadDelay)
+import           Control.Concurrent.Async
+
+import           Control.Monad (forever, unless, when)
+import           Control.Monad.Class.MonadSTM
+import           Control.Monad.Class.MonadThrow
+import           Control.Monad.Class.MonadTime
+import           Control.Monad.Class.MonadTimer
+import           Control.Tracer
+import Data.Set (Set)
+import qualified Data.Set as Set
+import           GHC.Stack
+import qualified Network.Socket as Socket
+import           Text.Printf
+
+import           Ouroboros.Network.Socket
+import           Ouroboros.Network.Subscription.Subscriber
+
+-- | Time to wait between connection attempts when we don't have any deltaQ info.
+defaultConnectionAttemptDelay :: DiffTime
+defaultConnectionAttemptDelay = 0.250 -- 250ms delay
+
+-- | Minimum time to wait between connection attempts.
+minConnectionAttemptDelay :: DiffTime
+minConnectionAttemptDelay = 0.010 -- 10ms delay
+
+-- | Maximum time to wait between connection attempts.
+maxConnectionAttemptDelay :: DiffTime
+maxConnectionAttemptDelay = 2 -- 2s delay
+
+-- | Minimum time to wait between ip reconnects
+ipRetryDelay :: DiffTime
+ipRetryDelay = 10 -- 10s delay
+
+sockAddrFamily
+    :: Socket.SockAddr
+    -> Socket.Family
+sockAddrFamily (Socket.SockAddrInet  _ _    ) = Socket.AF_INET
+sockAddrFamily (Socket.SockAddrInet6 _ _ _ _) = Socket.AF_INET6
+sockAddrFamily _                              = Socket.AF_UNSPEC
+
+data ConnectResult = ConnectSuccess
+                   | ConnectSuccessLast
+                   | ConnectFail
+                   deriving (Eq, Ord, Show)
+
+subscribeTo
+    :: HasCallStack
+    => ConnectionTable
+    -> TVar IO (Set ThreadId)
```

Comment:
Make a ticket for us to come back to this: we can probably either use async, or we can steal the thread registry from the consensus layer and use that.